client/components/app.jsx

```diff
@@ -0,0 +1,136 @@
+        '18%': 0.18,
+        '19%': 0.19,
+        '20%': 0.20,
+      },
+      tipSelected: null,
+      checkTotal: '',
+      checkTotalSet: false,
+      tipTotal: null,
+      grandTotal: null,
+      fateDecides: false,
+    };
+    this.onTotalCostChange = this.onTotalCostChange.bind(this);
+    this.onSetClick = this.onSetClick.bind(this);
+    this.calculateGrandTotal = this.calculateGrandtotal.bind(this);
+    this.handleTipPercentageClick = this.handleTipPercentageClick.bind(this);
+    this.handleReset = this.handleReset.bind(this);
+    this.letFateDecide = this.letFateDecide.bind(this);
+  }
+
+  onTotalCostChange(e) {
+    const name = e.target.name, value = e.target.value;
+    this.setState({ [name]: value });
+  }
+
+  onSetClick() {
+    this.setState({ checkTotalSet: !this.state.checkTotalSet }, () => {
+      if (this.state.fateDecides) {
+        const tipArr = this.state.tipSelected.split('');
+        tipArr.pop();
+        const tip = '0.' + tipArr.join();
+        this.calculateGrandTotal(tip)
+      }
+      else this.calculateGrandTotal(this.state.tipSelected)
+    })
+  }
+
+  handleReset() {
+    this.setState({ checkTotalSet: !this.state.checkTotalSet, tipTotal: null, grandTotal: null, checkTotal: '', tipSelected: null, fateDecides: false })
+  }
+
+  handleTipPercentageClick(e) {
+    if (this.state.fateDecides) {
+      const randomTip = Math.round(Math.random() * (40 - 15) + 15);
+      const percentage = '0.' + randomTip;
+      const tipSelected = percentage.split('.')[1] + '%';
+      this.setState({ tipSelected }, () => this.calculateGrandTotal(percentage))
+    } else {
+      const tipSelected = e.target.value;
+      this.setState({ tipSelected }, () => this.calculateGrandtotal(tipSelected));
+    }
+  }
+
+  letFateDecide(e) {
+    this.setState({ fateDecides: true }, () => this.handleTipPercentageClick())
+  }
+
+  calculateGrandtotal(tip) {
+    const { checkTotal, checkTotalSet, tipPercentage, fateDecides } = this.state;
+    if (!fateDecides) {
+      for (const key in tipPercentage) {
+        if (key === tip) {
+          const tipTotal = (tipPercentage[key] * checkTotal).toFixed(2);
+          const grandTotal = parseFloat(tipTotal) + parseFloat(checkTotal);
+          this.setState({ tipTotal, grandTotal });
+        }
+      }
+    } else {
+      const tipTotal = (parseFloat(tip) * checkTotal).toFixed(2);
+      const grandTotal = (parseFloat(tipTotal) + parseFloat(checkTotal)).toFixed(2);
+      this.setState({ tipTotal, grandTotal });
+      return;
+    }
+  }
+
+  render() {
+    const { tipPercentage, checkTotal, checkTotalSet, tipTotal, grandTotal, tipSelected } = this.state;
+    return (
+      <>
+        <Header title="TIP CALCULATOR" />
+        <main className="calculator-container container">
+          <section className="row">
+            <div className="col-12 d-flex">
+              CHECK TOTAL: {checkTotalSet ? <i style={{ marginLeft: '0.5em' }}>${checkTotal}</i> : null}
```

Comment:
I find JSX to be a bit complex inherently (a lot of mixed syntax), so consider keeping all logic out of the JSX and pulled away into functions. Name the function well, and the reader knows what it will accomplish. If the reader decides they want more details on how it works, they can find the function declaration to get more info.